File: Cpp/CXR_classify/labeler.cpp
#include "labeler.h"

Labeler::Labeler(ConfigHandler * configHandler, DatabaseHandler * dbHandler) : Runnable(configHandler, dbHandler)
{
    Labeler::labelTableName = configHandler->getTableName("label");
}

void Labeler::run()
{
    queryImageList();
    emit attemptUpdateText("Please manually label images");
    emit attemptUpdateProBarBounds(0, expected_num_files);

    dbHandler->addTableToDb(configHandler->getColumnsInfoPath(), "labels", labelTableName);

    displayNextImage();

    while (count < expected_num_files) {
        ;
    }

    emit attemptUpdateText("Image labeling complete");
    emit finished();
}

void Labeler::frontal()
{
    storeLabel("F");
    count++;
    displayNextImage();
}

void Labeler::lateral()
{
    storeLabel("L");
    count++;
    displayNextImage();
}

void Labeler::displayNextImage()
{
    emit attemptUpdateText("Image count: " + QString::number(count));
    emit attemptUpdateProBarValue(count);

    if (count < expected_num_files) {
        const char * filePath = record["file_path"].c_str();
        DicomImage * dcmImage = new DicomImage(filePath);
        uchar * pixelData = (uchar *) (dcmImage->getOutputData(8));

        int height = dcmImage->getHeight();
        int width = dcmImage->getWidth();

        cv::Mat originalImage(height, width, CV_8U, pixelData);
        cv::Mat imageSquare(height, width, CV_8U);
        cv::resize(originalImage, imageSquare, cv::Size(300, 300), 0, 0, cv::INTER_AREA);

        QImage qImage(imageSquare.data, 300, 300, QImage::Format_Grayscale8);
        QPixmap pixmap = QPixmap::fromImage(qImage);

        emit attemptUpdateImage(pixmap);
    }
}

void Labeler::storeLabel(std::string decision)
{
    std::string filePath = record["file_path"].c_str();
    std::string fileName = filePath.substr(filePath.find_last_of("/") + 1);
    std::string sqlQuery = "INSERT INTO " + labelTableName + "  (file_name, file_path, image_view) VALUES ('" + fileName + "', '" +
            filePath + "', '" + decision + "');";
    try
    {
        // Start a transaction
        pqxx::work storeTransaction(*connection);
        storeTransaction.exec(sqlQuery);
        storeTransaction.commit();
    }
    catch (std::exception const &e)
    {
        std::cerr << e.what() << std::endl;
    }

    record++;
}

void Labeler::queryImageList()
{
    std::string sqlQuery = "SELECT file_path, bits_stored FROM " + configHandler->getTableName("metadata") + " ORDER BY file_path;";
    try
    {
        // Connect to the database
        connection = dbHandler->openConnection();

        // Start a transaction
        pqxx::work imageListTransaction(*connection);

        // Execute query
        imageList = imageListTransaction.exec(sqlQuery);
        record = imageList.begin();

        imageListTransaction.commit();
    }
    catch (std::exception const &e)
    {
        // log "error"
    }
}

File: Python/DesktopApp/labeler.py
import logging
import os
import psycopg2
from PyQt5.QtCore import pyqtSignal, pyqtSlot
from stage import Stage

class Labeler(Stage):
    """Class used to assist in labeling the data."""
    attemptUpdateImage = pyqtSignal(object)

    def __init__(self, configHandler, dbHandler):
        Stage.__init__(self, configHandler, dbHandler)

        self.count = 0
        self.record = None
        self.cursor = None

    @pyqtSlot()
    def startLabeler(self):
        """Displays the content into the window."""
        logging.debug('Filling window')
        self.query_image_list()
        
        self.attemptUpdateText.emit("Please manually label images")
        self.attemptUpdateProBarBounds.emit(0, self.expected_num_files)
        self.attemptUpdateProBarValue.emit(0)

        self.dbHandler.add_table_to_db(self.configHandler.getTableName('label'), self.configHandler.getColumnsInfoPath(), 'labels')
        
        self.display_next_image()

    def frontal(self):
        logging.debug('Front')
        self.store_label('F')
        self.display_next_image()

    def lateral(self):
        logging.debug('Lateral')
        self.store_label('L')
        self.display_next_image()

    def display_next_image(self):
        logging.debug('Displaying next image')
        self.record = self.cursor.fetchone()

        self.attemptUpdateProBarValue.emit(self.dbHandler.count_records(self.configHandler.getTableName('label')))

        if self.record is None:
            logging.info('End of query')
            self.attemptUpdateText.emit("Image labeling complete")
            self.finished.emit()
        else:
            logging.debug('Image Count: ' + str(self.count))
            if self.count > 0:
                self.attemptUpdateText.emit('Image Count: ' + str(self.count))
            self.attemptUpdateImage.emit(self.record)
            self.count += 1
    
    def query_image_list(self):
        logging.debug('Getting the image list')
        sql_query = 'SELECT file_path, bits_stored FROM ' + self.configHandler.getTableName("metadata") + ' ORDER BY file_path;'
        self.cursor = self.dbHandler.openCursor(self.dbHandler.connection)
        self.dbHandler.executeQuery(self.cursor, sql_query)

    def store_label(self, decision):
        logging.debug('Storing label')
        sql_query = 'INSERT INTO ' + self.configHandler.getTableName("label") + ' (file_name, file_path, image_view) VALUES (\'' + self.record['file_path'].split(os.sep)[-1] + '\', \'' + self.record['file_path'] + '\', \'' + decision + '\');'
        cursor = self.dbHandler.openCursor(self.dbHandler.connection)
        self.dbHandler.executeQuery(cursor, sql_query)
